// @sap/approuter/lib/backend-request/options.js
'use strict';

const http = require('http');
const https = require('https');
const agents = require('./agents');
const backendHeaders = require('./headers');
const urlUtils = require('../utils/url-utils');
const traceUtil = require('../utils/trace-util');
const stripClientCookies = require('../middleware/strip-client-cookies');
const tough = require('tough-cookie');
const passportUtils = require('../passport/utils');
const _ = require('lodash');
const destinationUtils = require('../utils/destination-utils');
const businessServiceUtils = require('../utils/business-service-utils');

exports.getLogoutRequest = function (req, accessToken, destination, xsAppDestConfig, session) {
  let method = xsAppDestConfig.logoutMethod || 'POST';
  let logoutUrl = urlUtils.join(destination.url, xsAppDestConfig.logoutPath);
  let headers = backendHeaders.getHeaders(req, accessToken, destination);
  module.exports.getBackendCookies(req,session, headers, destination);
  return getRequestImpl(method, urlUtils.parse(logoutUrl), headers, destination, null,session);
};

exports.getBackendCookies = function(req,session, headers, destination){
  if (session && session._cookieJar) {
    if (req){
      stripClientCookies(req, req.app.get('cookieName'));
    }
    let jar = _.cloneDeep(session._cookieJar); // see tough-cookie issue #59
    tough.CookieJar.deserialize(jar, function (err, cookieJar) {
      if (err) {
        return;
      }
      cookieJar.getCookieString(destination.url, function(err, cookies){
        if (err) {
          return;
        }
        if (cookies) {
          headers.cookie = headers.cookie ? headers.cookie + '; ' + cookies : cookies;
        }
      });
    });
  }
};

exports.getRequest = function (req, accessToken) {
  let headers = backendHeaders.getHeaders(req, accessToken);
  let logger = req.logger;
  return getRequestImpl(req.method, req.internalUrl, headers, req.internalUrl.destination, logger, req.session);
};

function getRequestImpl(method, parsedUrl, headers, destination, logger, session) {
  const routeCredentials = parsedUrl.route && parsedUrl.route.credentials;
  const isMtls = routeCredentials && businessServiceUtils.getGrantType(routeCredentials) === businessServiceUtils.MTLS
      && routeCredentials.cert && routeCredentials.key;
  if (isMtls) {
    destination = Object.assign({}, destination, { strictConnectionMode: true });
  }

  let proxyUri = exports.buildProxyUri(destination);
  let opts = {
    method: method,
    protocol: parsedUrl.protocol,
    hostname: parsedUrl.hostname,
    port: parsedUrl.port,
    path: parsedUrl.path,
    agent: agents.get(parsedUrl.protocol, proxyUri, destination),
    rejectUnauthorized: destinationUtils.rejectUnauthorized(destination),
    headers: headers
  };

  if (isMtls) {
    opts.cert = routeCredentials.cert;
    opts.key = routeCredentials.key;
  } else if (destination && destination.cert && destination.key) {
    opts.cert = destination.cert;
    opts.key = destination.key;
  } else if (destination && destination.forwardAuthCertificates){
    passportUtils.getAuthCertificates(session, destination, opts);
  }

  if (headers.host) { // So socket does not use host header as target (allows differing host header)
    opts['servername'] = parsedUrl.hostname;
  }

  // http-proxy-agent v7 builds the proxy request-line absolute URL using req.getHeader('host')
  // rather than opts.hostname. For OnPremise destinations with preserveHostHeader=true, this
  // causes the preserved client host (e.g. myapp.cfapps.eu10.hana.ondemand.com) to appear in
  // the request line instead of the destination virtual host, breaking Cloud Connector ACL matching.
  // This applies to both normal requests (getRequest) and logout requests (getLogoutRequest):
  // in both cases the proxy request-line must use the destination virtual hostname for CC ACL matching.
  // Override headers.host to the destination hostname so the proxy request-line is correct.
  if (destination && destination.proxyType === 'OnPremise' && destination.preserveHostHeader && headers.host && parsedUrl.hostname) {
    opts.headers = Object.assign({}, headers, { host: parsedUrl.hostname });
  }

  let useSSL = /^https/i.test(opts.protocol);
  if (useSSL && opts.agent.options && opts.agent.options.ca) { // fixes leaking ssl sockets in node
    opts.ca = opts.agent.options.ca;
  }
  opts = destinationUtils.getHttpRequestOptions(destination, opts);
  if (logger) {
    traceUtil.traceBackendRequest(logger, parsedUrl.href, opts, destination);
  }
  return useSSL ? https.request.bind(undefined, opts) : http.request.bind(undefined, opts);
}

exports.buildProxyUri = function (destination) {
  let proxyHost = destination.proxyHost;
  let proxyPort = destination.proxyPort;
  if (!proxyHost) {
    return;
  }
  let hasProtocol = /^https?:\/\//i.test(proxyHost);
  if (hasProtocol) {
    return proxyHost + ':' + proxyPort;
  }
  return 'http://' + proxyHost + ':' + proxyPort;
};
