// @sap/approuter/lib/passport/oauth2.js
/* eslint-disable camelcase */

'use strict';

const _ = require('lodash');
const assert = require('assert');
const querystring = require('querystring');
const VError = require('verror').VError;
const traceUtils = require('../utils/trace-util');
const cookieUtils = require('../utils/cookie-utils');
const passportUtils = require('./utils');
const state = require('./state');
const requestTraceEnabled = require('../utils/trace-util').requestTraceEnabled;
const pathUtil = require('../utils/path-util');
const pkceUtils = require('../utils/pkce-utils');

module.exports.OAuth2 = OAuth2;

function OAuth2(options) {
  ['clientid', 'tokenURL'].forEach(function (element) {
    assert(options[element], 'OAuth2 requires "' + element + '" option');
  });
  assert(options.clientsecret || options.certificate, 'OAuth2 requires clientsecret or certificate');
  this._options = _.clone(options);
}

OAuth2.prototype.getCodeAuthorizationUrl = async function (req, authorizationURL, callbackURL) {
  const authURLQuery = {
    'response_type': 'code',
    'client_id': this._options.clientid,
    'redirect_uri': callbackURL
  };
  if (req.internalUrl && req.internalUrl.route && req.internalUrl.route.identityProvider) {
    authURLQuery.idp = req.internalUrl.route.identityProvider;
  }
  if (this._options.scope){ // IAS
    authURLQuery.scope = this._options.scope;
    authURLQuery.app_tid = this._options.apptid;
  }

  // State parameter for CSRF protection
  if (pathUtil.isStateProtectionEnabled(req) && cookieUtils.isBrowserRequest(req) && req.state) {
    authURLQuery['state'] =  req.state;
    state.setCacheStateValue(req.state);

    // PKCE enhancement - verifier keyed by session cookie (secret), not state (public)
    const pkceEnabled = req.routerConfig && req.routerConfig.pkceEnabled;
    if (pkceEnabled && req.preAuthSessionId) {
      // Generate verifier and challenge
      const pkce = pkceUtils.generatePkceChallenge();

      // Store verifier in STATE_CACHE using SESSION COOKIE value (secret)
      // NOT the state parameter (public) - this prevents URL-based attacks
      state.setCodeVerifier(req.preAuthSessionId, pkce.codeVerifier);

      // Add PKCE parameters to authorization URL
      authURLQuery['code_challenge'] = pkce.codeChallenge;
      authURLQuery['code_challenge_method'] = 'S256';

      const logger = require('../utils/logger');
      const tracer = logger.getTracer(__filename);
      tracer.debug('PKCE enabled for authorization request, state: %s', req.state);
    }
  }
  return authorizationURL + '?' + querystring.stringify(authURLQuery);
};

OAuth2.prototype.getOAuthAccessToken = function (params, loggingContext, callback) {
  const tracer = loggingContext.getTracer(__filename);
  const logger = loggingContext.getLogger('/request/uaa');
  if (tracer.isEnabled('debug')) {
    // Mask code_verifier in logs
    const logParams = Object.assign({}, params);
    if (logParams.code_verifier) {
      logParams.code_verifier = '***MASKED***';
    }
    tracer.debug('requesting UAA at %s: parameters %s', this._options.tokenURL, traceUtils.stringifyQueryParameters(logParams));
  }
  let requestStart = Date.now();
  params.client_id = this._options.clientid;
  if (this._options.scope === 'openid') { // IAS
    params.app_tid = this._options.apptid;
  }
  let requestOptions = {
    headers: this._options.customHeaders,
    url: this._options.tokenURL,
  };

  if (this._options.certificate){
    requestOptions.cert = this._options.certificate;
    requestOptions.key  = this._options.key;
  } else {
    params.client_secret = this._options.clientsecret;
    if (this._options.scope === 'openid'){ // IAS
      requestOptions.headers.Authorization = 'Basic ' + Buffer
        .from(this._options.clientid + ':' + this._options.clientsecret)
        .toString('base64');
    }
  }
  requestOptions.form = params;

  passportUtils.callUaa(requestOptions, null, function (err, uaaResponse) {
    if (err) {
      if (requestTraceEnabled) {
        logger.info('Request to authentication service at %s failed in %d ms, error: %s', requestOptions.url, Date.now() - requestStart, err.message);
      }
      let error = new VError('request to authentication service at %s failed, error: %s, ', requestOptions.url, err.message);
      error.status = err.status;
      return callback(error);
    }
    callback(null, uaaResponse);
  });
};
