// @sap/approuter/lib/passport/state.js
'use strict';

const NodeCache = require('node-cache');
const {getValidatedEnvInt} = require('../utils/env-utils');

const stdTTL              = getValidatedEnvInt('STATE_CACHE', 180);
const CHECK_PERIOD         = stdTTL + 20;
const STATE_CACHE          = new NodeCache({stdTTL: stdTTL, checkperiod: CHECK_PERIOD});

function setCacheStateValue(state) {
  STATE_CACHE.set(state, state);
}

function getCacheStateValue(state) {
  return STATE_CACHE.get(state);
}

/**
 * Delete a state cache value after use.
 *
 * @param {string} value - The cache key to delete
 */
function deleteCacheStateValue(value) {
  if (!value) {
    return;
  }
  STATE_CACHE.del(value);
}

/**
 * Store mapping from state parameter to session cookie value.
 * This allows validation that the callback request comes from the same browser
 * that initiated the OAuth flow, without exposing the cookie value in the URL.
 *
 * @param {string} state - The state parameter value (public, in URL)
 * @param {string} cookieValue - The session cookie value (secret, HttpOnly)
 */
function setStateToCookieMapping(state, cookieValue) {
  if (!state || !cookieValue) {
    throw new Error('State and cookie value are required');
  }
  STATE_CACHE.set(`state-cookie:${state}`, cookieValue);
}

/**
 * Retrieve the expected cookie value for a given state parameter.
 *
 * @param {string} state - The state parameter value
 * @returns {string|undefined} The expected cookie value, or undefined if not found
 */
function getStateToCookieMapping(state) {
  if (!state) {
    return undefined;
  }
  return STATE_CACHE.get(`state-cookie:${state}`);
}

/**
 * Delete state-to-cookie mapping after use.
 *
 * @param {string} state - The state parameter value
 */
function deleteStateToCookieMapping(state) {
  if (!state) {
    return;
  }
  STATE_CACHE.del(`state-cookie:${state}`);
}

/**
 * Check if a cookie value corresponds to an active OAuth flow.
 * Used to filter out pre-auth session cookies from stale cookie cleanup.
 *
 * @param {string} cookie - Cookie string in format "name=value"
 * @returns {boolean} True if cookie is associated with an active OAuth flow
 */
function isActiveStateCookie(cookie) {
  try {
    const cookieValue = cookie.trim().split('=')[1];
    return STATE_CACHE.has(cookieValue);
  } catch (_ignored) {
    return false;
  }
}

/**
 * Store PKCE code verifier associated with session cookie value.
 * The verifier is keyed by the COOKIE value (secret), not the state (public).
 * This prevents attackers who can read URLs from retrieving the verifier.
 *
 * @param {string} sessionId - The session cookie value (secret)
 * @param {string} codeVerifier - The PKCE code verifier to store
 * @throws {Error} If sessionId or codeVerifier is missing
 */
function setCodeVerifier(sessionId, codeVerifier) {
  if (!sessionId || !codeVerifier) {
    throw new Error('Session ID and code verifier are required');
  }
  STATE_CACHE.set(`pkce:${sessionId}`, codeVerifier);
}

/**
 * Retrieve PKCE code verifier for session cookie value.
 *
 * @param {string} sessionId - The session cookie value
 * @returns {string|undefined} The code verifier, or undefined if not found
 */
function getCodeVerifier(sessionId) {
  if (!sessionId) {
    return undefined;
  }
  return STATE_CACHE.get(`pkce:${sessionId}`);
}

/**
 * Delete PKCE code verifier after use.
 *
 * @param {string} sessionId - The session cookie value
 */
function deleteCodeVerifier(sessionId) {
  if (!sessionId) {
    return;
  }
  STATE_CACHE.del(`pkce:${sessionId}`);
}

module.exports = {
  setCacheStateValue,
  getCacheStateValue,
  deleteCacheStateValue,
  isActiveStateCookie,
  setStateToCookieMapping,
  getStateToCookieMapping,
  deleteStateToCookieMapping,
  setCodeVerifier,
  getCodeVerifier,
  deleteCodeVerifier
};
