// @sap/approuter/lib/configuration/env-config.js
'use strict';

const _ = require('lodash');
const logger = require('../utils/logger').getLogger('/Configuration');
const path = require('path');
const safeRegex = require('safe-regex');
const VError = require('verror').VError;
const xsenv = require('@sap/xsenv');
const configurationUtils = require('../utils/configuration-utils');
let validators;
const {normalizeDestinationProperties, adjustDestinationProperties} = require('../utils/destination-normalization');

function getValidators() {
  if (!validators) {
    validators = require('./validators');
  }
  return validators;
}

exports.load = load;
exports.loadDestinations = loadDestinations;
exports.loadJsonVar = loadJsonVar;

let _mainAppConfig;

Object.defineProperty(exports, 'mainAppConfig', {
  get: function() {
    if (!_mainAppConfig) {
      throw new VError('Main app config not loaded');
    }
    return _mainAppConfig;
  }
});

exports.resetMainAppConfig = function() {
  _mainAppConfig = undefined;
};

function load(workingDir) {
  const filePath = path.join(workingDir || '', 'default-env.json');
  xsenv.loadEnv(filePath);
  let envConfig = {};
  envConfig.destinations = loadDestinations();
  envConfig.additionalHeaders = loadAdditionalHeaders();
  envConfig.plugins = loadPlugins(envConfig.destinations);
  envConfig.clickjackCheckWhitelist = loadClickjackWhitelist();
  envConfig.wsAllowedOrigins = loadWebSocketWhitelist();
  envConfig.secureSessionCookie = resolveSessionCookieSecureMode();
  envConfig.cors = loadCors();
  envConfig.cookies = loadCookies();
  envConfig.extSessionMgt = loadExternalSessionManager();
  envConfig.circuitBreakerConf = loadCircuitBreakerConfiguration();

  const grouped = loadGroupedConfigurations();
  _mainAppConfig = _.merge(envConfig, grouped);
  return _mainAppConfig;
}

function loadExternalSessionManager() {
  const extSessionMgt = loadJsonVar('EXT_SESSION_MGT');
  if (!extSessionMgt) {
    return null;
  }
  getValidators().validateSessionStoreInstance(extSessionMgt);
  return extSessionMgt;
}

function loadCircuitBreakerConfiguration() {
  const circuitBrakerConf = loadJsonVar('CIRCUIT_BREAKER_CONF');
  if (!circuitBrakerConf) {
    return null;
  }
  getValidators().validateCircuitBreakerConf(circuitBrakerConf);
  return circuitBrakerConf;
}

function loadCookies() {
  const cookies = loadJsonVar('COOKIES');
  if (!cookies) {
    logger.info('No COOKIES environment variable');
    return null;
  }
  getValidators().validateCookies(cookies);
  if (cookies.Partitioned){
    cookies.Partitioned.supportedPartitionAgents = configurationUtils.constructRegExp(cookies.Partitioned.supportedPartitionAgents);
    cookies.Partitioned.unsupportedPartitionAgents = configurationUtils.constructRegExp(cookies.Partitioned.unsupportedPartitionAgents);
    if (!safeRegex(cookies.Partitioned.supportedPartitionAgents)) {
      logger.warning(`supportedPartitionAgents regex ${cookies.Partitioned.supportedPartitionAgents} is vulnerable to ReDoS attacks`);
    }
    if (!safeRegex(cookies.Partitioned.unsupportedPartitionAgents)) {
      logger.warning(`unsupportedPartitionAgents regex ${cookies.Partitioned.unsupportedPartitionAgents} is vulnerable to ReDoS attacks`);
    }
  }
  return cookies;
}


function loadCors() {
  const cors = loadJsonVar('CORS');
  if (!cors) {
    return;
  }
  getValidators().validateCors(cors);
  cors.forEach(function (corsConfig) {
    if (corsConfig.hostPattern) {
      corsConfig.hostPattern = configurationUtils.constructRegExp(corsConfig.hostPattern);
      if (!safeRegex(corsConfig.hostPattern)) {
        logger.warning('Cors configuration with hostPattern ', corsConfig.hostPattern, ' is vulnerable to ReDoS attacks');
      }
    }
    corsConfig.uriPattern = configurationUtils.constructRegExp(corsConfig.uriPattern);
    if (!safeRegex(corsConfig.uriPattern)) {
      logger.warning('Cors configuration with uriPattern ', corsConfig.uriPattern, ' is vulnerable to ReDoS attacks');
    }
  });
  return cors;
}

function loadDestinations(destinations) {
  destinations = destinations || loadJsonVar('destinations');
  if (!destinations) {
    destinations = {};
    logger.info('Using empty destinations to run');
    return destinations;
  }
  normalizeDestinationProperties(destinations);
  getValidators().validateDestinations(destinations);
  adjustDestinationProperties(destinations);
  return _.keyBy(destinations, 'name');
}

function loadAdditionalHeaders() {
  let additionalHeaders = loadJsonVar('httpHeaders');
  if (!additionalHeaders) {
    return [];
  }
  getValidators().validateHeaders(additionalHeaders);
  additionalHeaders = _.map(additionalHeaders, function (header) {
    const normalizedHeader = {};
    _.forEach(header, function (value, key) {
      normalizedHeader[key.toLowerCase()] = value;
    });
    return normalizedHeader;
  });
  return additionalHeaders;
}

function loadPlugins(destinations) {
  const plugins = loadJsonVar('plugins');
  if (!plugins) {
    return null;
  }
  getValidators().validatePlugins(plugins, destinations);
  plugins.forEach(function (currentPlugin) {
    currentPlugin.source = configurationUtils.constructRegExp(currentPlugin.source);
  });
  return plugins;
}

function loadOutgoingRequestOptions() {
  const outgoingRequestOptionsEnv = loadJsonVar('OUTGOING_REQUEST_OPTIONS') || {};
  const {DEFAULT_OUTGOING_REQUEST_OPTIONS} = require('./outgoing-request-options');
  Object.keys(outgoingRequestOptionsEnv).forEach(function (key) {
    if (!(key in DEFAULT_OUTGOING_REQUEST_OPTIONS)) {
      throw new VError('Invalid outgoing request option: "%s". Available options are "%s"', key, Object.keys(DEFAULT_OUTGOING_REQUEST_OPTIONS).join(', '));
    }
  });
  const defaults = _.cloneDeep(DEFAULT_OUTGOING_REQUEST_OPTIONS);
  return _.merge(defaults, outgoingRequestOptionsEnv);
}

function loadClickjackWhitelist() {
  return loadOriginsWhitelist('CJ_PROTECT_WHITELIST');
}

function loadWebSocketWhitelist() {
  return loadOriginsWhitelist('WS_ALLOWED_ORIGINS');
}

function loadOriginsWhitelist(envVar) {
  const whitelist = loadJsonVar(envVar);

  if (!whitelist) {
    return null;
  }
  getValidators().validateWhitelist(whitelist);
  whitelist.forEach(function (item) {
    if (typeof item.port === 'string') {
      item.port = parseInt(item.port);
    }
  });
  return whitelist;
}

function resolveSessionCookieSecureMode() {
  switch (process.env.SECURE_SESSION_COOKIE) {
  case 'true':
    return true;
  case 'false':
    return false;
  default:
    return 'auto';
  }
}

function loadGroupedConfigurations() {
  let grouped = {
    sessionTimeout: loadJsonVar('SESSION_TIMEOUT'),
    incomingConnectionTimeout: loadJsonVar('INCOMING_CONNECTION_TIMEOUT'),
    incomingRequestTimeout: loadJsonVar('INCOMING_REQUEST_TIMEOUT'),
    jwtRefresh: loadJsonVar('JWT_REFRESH'),
    tenantHostPattern: process.env['TENANT_HOST_PATTERN'],
    destinationHostPattern: process.env['DESTINATION_HOST_PATTERN'],
    compression: loadJsonVar('COMPRESSION'),
    externalReverseProxy: loadJsonVar('EXTERNAL_REVERSE_PROXY'),
    storeSessionCookiesInExternalSessionStore: loadJsonVar('STORE_SESSION_COOKIES_IN_EXTERNAL_SESSION_STORE'),
    directRoutingUriPatterns: loadJsonVar('DIRECT_ROUTING_URI_PATTERNS'),
    cfNodeJsLoggingLevel: process.env['CF_NODEJS_LOGGING_LEVEL'] || 'error',
    http2Support: loadJsonVar('HTTP2_SUPPORT'),
    serverKeepAlive: loadJsonVar('SERVER_KEEP_ALIVE'),
    minimumTokenValidity: loadJsonVar('MINIMUM_TOKEN_VALIDITY'),
    dynamicIdentityProvider: loadJsonVar('DYNAMIC_IDENTITY_PROVIDER'),
    stateParameterSecret: process.env['STATE_PARAMETER_SECRET'],
    excludedStateHandlingSubdomains: loadJsonVar('EXCLUDED_STATE_HANDLING_SUBDOMAINS'),
    enableXForwardHostValidation: loadJsonVar('ENABLE_X_FORWARDED_HOST_VALIDATION'),
    enableCSPheaders: loadJsonVar('ENABLE_FRAME_ANCESTORS_CSP_HEADERS'),
    ownSapCloudService: loadJsonVar('OWN_SAP_CLOUD_SERVICE'),
    appsCacheExpiration: loadJsonVar('APPS_CACHE_EXPIRATION'),
    cookieBackwardCompatibility: loadJsonVar('COOKIE_BACKWARD_COMPATIBILITY'),
    saasApprouterExcludedPath: loadJsonVar('SAAS_APPROUTER_EXCLUDED_PATH'),
    backChannelLogoutUri: process.env['BACK_CHANNEL_LOGOUT_URI'],
    checkXForwardedHostInLoginCallback: loadJsonVar('CHECK_X_FORWARDED_HOST_IN_LOGIN_CALLBACK'),
    lockTTL: loadJsonVar('APPS_CACHE_LOCK_TTL_MS') || 20000,
    pollingInterval: loadJsonVar('APPS_CACHE_POLL_INTERVAL_MS') || 1000,
    maxWaitTime: loadJsonVar('APPS_CACHE_MAX_WAIT_MS') || 120000,
    outgoingRequestOptions: loadOutgoingRequestOptions(),
    returnErrorSource: loadJsonVar('RETURN_ERROR_SOURCE')
  };

  grouped = _.pickBy(grouped, function (value) {
    return (value || value === '' || value === 0);
  });

  if (!grouped.hasOwnProperty('jwtRefresh')) {
    grouped.jwtRefresh = 5; // 5 is the default JWT_REFRESH
  }
  const sendXFrameOptions = loadJsonVar('SEND_XFRAMEOPTIONS');
  grouped.sendXFrameOptions = (typeof sendXFrameOptions !== 'undefined') ? sendXFrameOptions : true;
  const pkceEnabled = loadJsonVar('PKCE_ENABLED');
  grouped.pkceEnabled = (typeof pkceEnabled !== 'undefined') ? pkceEnabled : true;
  getValidators().validateEnvironmentSettings(grouped);

  if (grouped.tenantHostPattern) {
    grouped.tenantHostPattern = configurationUtils.constructRegExp(grouped.tenantHostPattern);
    if (!safeRegex(grouped.tenantHostPattern)) {
      logger.warning('TENANT_HOST_PATTERN is vulnerable to ReDoS attacks');
    }
  }

  if (grouped.destinationHostPattern) {
    grouped.destinationHostPattern = configurationUtils.constructRegExp(grouped.destinationHostPattern);
    if (!safeRegex(grouped.destinationHostPattern)) {
      logger.warning('DESTINATION_HOST_PATTERN is vulnerable to ReDoS attacks');
    }
  }

  if (grouped.directRoutingUriPatterns) {
    grouped.directRoutingUriPatterns.forEach(function (currentDirectPattern) {
      if (getValidators().checkStringOrRegex(currentDirectPattern)) {
        return;
      }
      throw new VError('Direct routing uri pattern %s must be a string or a regular expression', currentDirectPattern.toString());
    });
  }

  return grouped;
}

function loadJsonVar(envVar) {
  if (envVar in process.env) {
    try {
      return JSON.parse(process.env[envVar]);
    } catch (e) {
      throw new VError(e, 'Invalid value for environment variable %s', envVar);
    }
  }
}
