// @sap/approuter/lib/middleware/access-token-handler.js
'use strict';

const jwtDecode = require('jwt-decode');
const crypto = require('crypto');
const tokenUtils = require('../utils/token-utils');
const headerUtils = require('../utils/header-util');
const vcapUtils = require('../utils/vcap-utils');
const {writeAuditLog} = require('../utils/audit-log');
const NodeCache           = require('node-cache');
const {getValidatedEnvInt} = require('../utils/env-utils');
const tokenCacheTTL = getValidatedEnvInt('TOKEN_CACHE_TTL', 3600); // default 1 hour
const tokenCache      = new NodeCache({stdTTL: tokenCacheTTL, checkperiod: tokenCacheTTL + 20, maxKeys: 10000}); // ~40 Mb max considering token size ~4kb

const X_APPROUTER_ACCESS_TOKEN = 'x-approuter-access-token';
const X_APPROUTER_AUTHORIZATION = 'x-approuter-authorization';

const IDENTITY = 'identity';
const ID_TOKEN = 'id_token';

module.exports = {
  hasAccessTokenHeader,
  handleToken,
  exchangeTokenCached
};

class ServiceToApprouterException extends Error {
  constructor({error: {message, status}, credentials, tokenDecoded}) {
    super(message);
    this.name = ServiceToApprouterException.name;
    this.error = {message, status};
    this.credentials = credentials || null;
    this.tokenDecoded = tokenDecoded || null;
  }
}

async function handleToken(req) {
  const tracer = req.loggingContext.getTracer(__filename);
  tracer.debug(`Incoming request contains ${X_APPROUTER_ACCESS_TOKEN} header - putting it inside ${X_APPROUTER_AUTHORIZATION} after extension`);
  const credentials = vcapUtils.getServiceCredentials({label: IDENTITY});
  if (process.env.IAS_PRIVATE_KEY){
    tracer.debug('Overriding raw key certificate from IAS_PRIVATE_KEY env');
    credentials.key = process.env.IAS_PRIVATE_KEY;
  }

  try {
    const weakConsumerToken = req.headers[X_APPROUTER_ACCESS_TOKEN];
    const extendedTokenValue = await exchangeWeakToken(req, weakConsumerToken, credentials, tracer);
    req.headers[X_APPROUTER_AUTHORIZATION] = `Bearer ${extendedTokenValue}`;
    delete req.headers[X_APPROUTER_ACCESS_TOKEN];
  } catch (e) {
    if (e instanceof ServiceToApprouterException) {
      const {error, credentials, tokenDecoded} = e;
      if (error.status === 401 || error.status === 403) {
        writeAuditLog(
          req,
          IDENTITY,
          error.message,
          error.status,
          tokenDecoded,
          credentials,
        );
      }
    }
    throw e;
  }
}

function hasAccessTokenHeader(req) {
  return req.headers && req.headers[X_APPROUTER_ACCESS_TOKEN];
}

// short-lived IAS/XSUAA token handling
async function exchangeWeakToken(req, weakTokenValue, credentials, tracer) {
  const {tokenValue, tokenDecoded} = getDecodedToken(weakTokenValue, X_APPROUTER_ACCESS_TOKEN);
  if (!tokenDecoded.jti) {
    tracer.error(`${X_APPROUTER_ACCESS_TOKEN} is invalid since it has no jti claim`);
    throw new ServiceToApprouterException({
      serviceName: tokenValue,
      error: {
        message: `${X_APPROUTER_ACCESS_TOKEN} is invalid`,
        status: 400
      }
    });
  }
  const correlationId = headerUtils.getCorrelationId(req);
  // eslint-disable-next-line camelcase
  const {id_token} = await exchangeTokenCached(
    tokenValue,
    tokenDecoded,
    correlationId,
    credentials,
    tracer,
  );
  // eslint-disable-next-line camelcase
  return id_token;
}

async function exchangeTokenCached(
  tokenValue,
  tokenDecoded,
  correlationId,
  credentials,
  tracer,
) {
  // Create a cache key based on the token value so man in the middle can't exploit jti with signed token
  const tokenKey = crypto.createHash('sha256').update(tokenValue).digest('hex');
  tracer.debug('exchangeToken cached start', tokenKey);
  if (tokenCache.has(tokenKey)) {
    const tokenResponse = tokenCache.get(tokenKey);
    if (tokenResponse.exp - (Date.now() / 1000) > 0) {
      tracer.debug(`Found token type ${tokenResponse.type} in cache, returning it`);
      return tokenResponse; // the same signature as the other return
    }
    tracer.debug(`Found expired token of type ${tokenResponse.type} in cache - deleting it`);
    tokenCache.del(tokenKey);
  }

  const tokenResponse = await tokenUtils.exchangeIdentityTokenByIssuer(tokenValue, tokenDecoded, correlationId, credentials);
  const tokenDecodedNew = jwtDecode(tokenResponse[ID_TOKEN]);
  const cacheTtl = Math.min(tokenCacheTTL, Math.abs(Math.floor(tokenDecodedNew.exp - (Date.now() / 1000))));
  tracer.debug(`exchangeToken successfully extended token of type ${IDENTITY}, caching it with ttl ${cacheTtl} seconds`);
  const result = {exp: tokenDecodedNew.exp, [ID_TOKEN]: tokenResponse[ID_TOKEN]};
  try {
    tokenCache.set(tokenKey, result, cacheTtl);
  } catch (e) {
    if (e.errorcode && e.errorcode === 'ECACHEFULL') {
      tracer.error(`Failed to set token in cache: ${e.message}`);
    } else {
      throw e;
    }
  }
  return result;
}

function getDecodedToken(auth, headerName) {
  let tokenDecoded;
  try {
    tokenDecoded = jwtDecode(auth);
  } catch (err) {
    throw new ServiceToApprouterException({
      error: {
        message: `Failed to decode ${headerName} token: ${err}`,
        status: 400
      }
    });
  }
  const tokens = auth.split(' ');
  const [type, tokenValue] = tokens.length === 2 ? tokens : [undefined, auth];
  return {type, tokenValue, tokenDecoded};
}
