// @sap/approuter/lib/utils/destination-normalization.js
'use strict';

const _ = require('lodash');
const {getServiceCredentials} = require('./vcap-utils');

function normalizeDestination(destination) {
  // IMPORTANT: Don't change normalize logic as it might break approuter extensions
  if (!destination) {
    return;
  }
  let keys = Object.keys(destination);
  let n = keys.length;
  let newobj = {};
  let key;
  while (n--) {
    key = keys[n];
    if (key === 'URL') {
      newobj[key.toLowerCase()] = destination[key];
    } else {
      let value = destination[key];
      key = key.replace('HTML5.', '');
      let newKey = key.substr(0, 1).toLowerCase() + key.substr(1);
      newobj[newKey] = value;
    }
  }
  return newobj;
}

function normalizeDestinationProperties(destinations) {
  if (!destinations) {
    return;
  }
  if (!Array.isArray(destinations)) {
    throw new Error('destinations type invalid, array is expected');
  }
  destinations.forEach(function (destination, i, destinations) {
    if (destinations[i].Authentication) { // if it destination from destinations service
      destinations[i] = normalizeDestination(destination);
      destinations[i].forwardAuthToken = destinations[i].forwardAuthToken === 'true' || destinations[i].forwardAuthToken === true;
      destinations[i].forwardAuthCertificates = destinations[i].forwardAuthCertificates === 'true' || destinations[i].forwardAuthCertificates === true;
      destinations[i].preserveHostHeader = destinations[i].preserveHostHeader === 'true' || destinations[i].preserveHostHeader === true;
      destinations[i].dynamicDestination = (destinations[i].hasOwnProperty('dynamicDestination') && destinations[i].dynamicDestination ? destinations[i].dynamicDestination.toLowerCase() === 'true' : false)  || destinations[i].dynamicDestination === true;
      destinations[i].trustAll = destinations[i].trustAll === 'true' || destinations[i].trustAll === true;
      destinations[i].strictConnectionMode = destinations[i].strictConnectionMode === 'true' || destinations[i].strictConnectionMode === true;
      if (destinations[i].hasOwnProperty('setXForwardedHeaders')) {
        destinations[i].setXForwardedHeaders = destinations[i].setXForwardedHeaders === 'true' || destinations[i].setXForwardedHeaders === true;
      }
      if (destinations[i].timeout && !isNaN(destinations[i].timeout)) {
        destinations[i].timeout = _.toSafeInteger(destinations[i].timeout);
      }
    }
  });
}

function adjustDestinationProperties(destinations) {
  destinations.forEach(function (destination) {
    if (destination.proxyType === 'OnPremise') {
      try {
        let credentials = getServiceCredentials({tag: 'connectivity'});
        destination['proxyHost'] = credentials.onpremise_proxy_host;
        destination['proxyPort'] = credentials.onpremise_proxy_port;
      } catch (_error) {
        throw new Error(`Destination "${destination.name}" with  ProxyType "OnPremise" but connectivity service is not bound.`);
      }
    }
    _.defaults(destination, {
      timeout: 30000
    });
  });
}

module.exports = {
  normalizeDestinationProperties,
  adjustDestinationProperties,
  normalizeDestination
};
