// @sap/approuter/lib/utils/request-utils.js
'use strict';
const _ = require('lodash');
const axios  = require('axios');
const axiosRetry = require('axios-retry').default;
const { wrapper }  = require('axios-cookiejar-support');
const querystring = require('querystring');
const https = require('https');
const tracer = require('./logger').getTracer(__filename);
const VError = require('verror').VError;
const envConfig = require('../configuration/env-config');

// support cookiejar
wrapper(axios);
axios.defaults.withCredentials = true;

// Configurable retry parameters
const RETRY_CONFIG = {
  retries: 3,                // Maximum number of retry attempts
  retryDelay: 1000,          // Base delay in milliseconds (will be multiplied by retry count)
  maxRetryDelay: 5000,       // Maximum delay between retries
  retryMultiplier: 2,        // Multiplier for exponential backoff
  retryNetworkErrors: true,  // Whether to retry network errors
  retry5xxErrors: [502]      // Whether to retry 5xx server errors
};

// convert request options to 'axios' format
const axiosRequestOptions = (method , request) => {
  const clonedRequestOpt = _.clone(request);
  clonedRequestOpt.method = method;
  clonedRequestOpt.transformResponse = function (res){   // prevent JSON.parse
    return res;
  };
  if (clonedRequestOpt.cert && clonedRequestOpt.key){
    const agentOptions = {
      cert:clonedRequestOpt.cert,
      key:clonedRequestOpt.key
    };
    if (https.globalAgent.options && https.globalAgent.options.ca) {
      agentOptions.ca = https.globalAgent.options.ca;
      tracer.debug('Using custom CA certificates for certificate authentication request');
    }
    clonedRequestOpt.httpsAgent = new https.Agent(agentOptions);
    delete clonedRequestOpt.cert;
    delete clonedRequestOpt.key;
  }
  if (clonedRequestOpt.form){
    if (!clonedRequestOpt.headers){
      clonedRequestOpt.headers = {};
    }
    clonedRequestOpt.headers['Content-Type'] = 'application/x-www-form-urlencoded';
    clonedRequestOpt.data = querystring.stringify(clonedRequestOpt.form);
    delete clonedRequestOpt.form;
  }
  if (clonedRequestOpt.auth) {
    clonedRequestOpt.auth.username = clonedRequestOpt.auth.user;
    clonedRequestOpt.auth.password = clonedRequestOpt.auth.pass;
    delete clonedRequestOpt.auth.user;
    delete  clonedRequestOpt.auth.pass;
  }
  if (clonedRequestOpt.body){
    clonedRequestOpt.data = clonedRequestOpt.body;
    delete clonedRequestOpt.body;
  }
  if (clonedRequestOpt.followRedirect === false){
    clonedRequestOpt.maxRedirects = 0 ;  // prevent redirect
    clonedRequestOpt.validateStatus = function (status) {
      return status >= 200 && status < 303; //  prevent Promise.reject when status code 302
    };
    delete clonedRequestOpt.followRedirect;
  }
  if (clonedRequestOpt.service) {
    clonedRequestOpt.timeout = getTimeoutByService(clonedRequestOpt.service);
    delete clonedRequestOpt.service;
  }
  return clonedRequestOpt;
};

const get = (requestOptions, callback) => {
  (async () => {
    const {error , response , body} = await axiosRequest('get',requestOptions);
    return callback(error ,response , body);
  })();
};

const post = (requestOptions,callback) => {
  (async () => {
    const {error , response , body} = await axiosRequest('post',requestOptions);
    return callback(error ,response , body);
  })();
};

const pipe = async (requestOptions, response, reject) => {
  requestOptions.responseType = 'stream';
  const resp = await axiosRequest('get',requestOptions);
  if (resp.error) {
    if (reject && typeof reject === 'function') {
      return reject(resp.error);
    }
    throw resp.error;
  }
  resp.body.pipe(response);
};

function defaultRetryCondition(error) {
  // If there's no response, it's a network error or even timeout (including axios)
  if (!error.response && RETRY_CONFIG.retryNetworkErrors) {
    return true;
  }

  // Only retry on 5xx errors if configured to do so
  return (
    Array.isArray(RETRY_CONFIG.retry5xxErrors) &&
    error.response &&
    RETRY_CONFIG.retry5xxErrors.includes(error.response.status)
  );
}

const axiosRequest = async (method, requestOptions) => {
  let error = null;
  let cache = requestOptions.cache;
  const options = axiosRequestOptions(method, requestOptions);

  // Check if retry is enabled for this request
  const enableRetry = requestOptions.enableRetry === true && !(process.env.DISABLE_REQUEST_RETRIES === 'true');

  let response;

  if (enableRetry) {
    const customRetryCondition = requestOptions.retryCondition;
    const customRetries = requestOptions.retries;
    // Create a new axios instance for this request
    const axiosInstance = axios.create();

    // Configure axios-retry for this instance
    axiosRetry(axiosInstance, {
      retries: customRetries || RETRY_CONFIG.retries,

      // Use exponential backoff with configurable parameters
      retryDelay: (retryCount) => {
        const delay = RETRY_CONFIG.retryDelay * Math.pow(RETRY_CONFIG.retryMultiplier, retryCount - 1);
        return Math.min(delay, RETRY_CONFIG.maxRetryDelay);
      },

      // Only retry on network errors or 5xx responses based on configuration
      retryCondition: customRetryCondition || module.exports.defaultRetryCondition
    });

    // Make the request with the retry-enabled instance
    try {
      response = await axiosInstance(options);
    } catch (err) {
      // the response in case of error is the error - all the logic farther based on it
      response = error = err;
    }
  } else {
    // Make the request without retries
    try {
      response = await axios(options);
    } catch (err) {
      // the response in case of error is the error - all the logic farther based on it
      response = error = err;
    }
  }

  if (response){
    response.statusCode = response.status;
    response.body = response.data;
    if (!response.statusCode){
      const statusCode = error && error.response && error.response.status;
      response.statusCode = statusCode;
    }
  }
  if (response && response.statusCode && (response.statusCode ===  401 || response.statusCode === 403) &&
      requestOptions.headers && requestOptions.headers.Authorization && requestOptions.headers.Authorization.startsWith('Bearer ')) {
    if (cache && cache.services && cache.serviceName) {
      delete cache.services[cache.serviceName].token;
      if (cache.tenant && cache.services[cache.serviceName][cache.tenant]) {
        delete cache.services[cache.serviceName][cache.tenant].token;
      }
    }
  }
  return {
    error : (error && error.response) ? null : error,
    response,
    body : response ? response.data : null,
  };
};

/*
  refer to lib/configuration/default/outgoung-request-options.json
*/
function getTimeoutByService(serviceName) {
  const timeout = envConfig.mainAppConfig.outgoingRequestOptions[serviceName].timeout;
  if (timeout === undefined) {
    throw new VError('Timeout for service %s is not configured', serviceName);
  }
  return timeout;
}

module.exports = {
  post,
  get,
  pipe,
  axiosRequestOptions,
  axiosRequest,
  defaultRetryCondition,
};
